include/eve/arch/riscv/tags.hpp

```diff
@@ -0,0 +1,115 @@
+//==================================================================================================
+/*
+  EVE - Expressive Vector Engine
+  Copyright : EVE Project Contributors
+  SPDX-License-Identifier: BSL-1.0
+*/
+//==================================================================================================
+#pragma once
+
+#include <eve/arch/cardinals.hpp>
+#include <eve/arch/cpu/tags.hpp>
+#include <eve/arch/riscv/predef.hpp>
+#include <eve/concept/scalar.hpp>
+#include <eve/detail/meta.hpp>
+
+#include <bit>
+
+namespace eve
+{
+//================================================================================================
+// ABI tags for all RISCV bits SIMD registers
+//================================================================================================
+template<std::size_t Size> struct rvv_abi_
+{
+  static constexpr std::size_t MaxLmul         = 8;
+  static constexpr std::size_t bits            = Size * MaxLmul;
+  static constexpr std::size_t bytes           = bits / 8;
+  static constexpr bool        is_wide_logical = false;
+
+  template<typename Type>
+  static constexpr bool is_full = ((Type::size() * sizeof(typename Type::value_type)) >= bytes);
+
+  template<typename Type>
+  static constexpr std::size_t fundamental_cardinal = Size / 8 / sizeof(Type);
+
+  template<typename Type>
+  static constexpr std::size_t expected_cardinal = Size / 8 / sizeof(Type) * MaxLmul;
+
+  template<typename Type> static consteval int get_min_frac_lmul()
+  {
+    auto type_size = sizeof(Type);
+    if( type_size == 1 ) return 8;
+    if( type_size == 2 ) return 4;
+    if( type_size == 4 ) return 2;
+    // for bigger types we can not use frac lmul.
+    // Return bigger than size of vector register value.
+    return Size + 1;
+  }
+
+  template<typename Type> static consteval int get_max_frac_lmul()
+  {
+    auto type_size = sizeof(Type);
+    if( type_size <= 4 ) return 2;
+    // for bigger types we can not use frac lmul.
+    // Return bigger than size of vector register value.
+    return Size + 1;
+  }
+
+  // return natural lmul if > 0, frac otherwise
+  template<typename Type> static consteval int getLMUL(size_t cardinal)
+  {
+    auto binary_size   = cardinal * sizeof(Type) * 8;
+    auto min_frac_lmul = get_min_frac_lmul<Type>();
+    auto max_frac_lmul = get_max_frac_lmul<Type>();
+
+    auto frac_lmul_border = Size / max_frac_lmul;
+    if( binary_size <= frac_lmul_border )
+    {
+      auto frac_lmul = Size / binary_size;
+      frac_lmul      = std::bit_floor(frac_lmul);
+      if( frac_lmul > min_frac_lmul ) frac_lmul = min_frac_lmul;
+      return -frac_lmul;
+    }
+    auto lmul = binary_size / Size;
+    if( binary_size % Size ) ++lmul;
+    return std::bit_ceil(lmul);
+  }
+
+  template<arithmetic_scalar_value T> using m1_fixed = fixed<Size / 8 / sizeof(T)>;
+};
+
+#ifdef __riscv_v_fixed_vlen
+struct riscv_rvv_dyn_ : rvv_abi_<__riscv_v_fixed_vlen>
+{};
+#else
+struct riscv_rvv_dyn_ : rvv_abi_<1>
```

Comment:
Currently, if vlen is not specified (that is no `-mrvv-vector-bits` is set), compilation results in an error. This happens because constexprs are unable to be evaluated properly. It seems that emulation does not work for RISC-V even without my changes. 

My idea was following: I wanted to try an set vlen to some non-sensible value in a hope that this will force template magic to enforce emulation. Now, I see that this attempt failed because of the following errors (again, this happens without my patches):

```
error: invalid operands to binary expression ('const ::eve::cpu_' and const asimd_')
...
```

For now added instead of this `#error` with description, for these compilation failures to be a little less cryptic. 

By the way, do you maybe have any recommendations on how to make emulation work?

Reply:
> I need more context. Does it fail when you try to compile without any changes? Where?

I see currently errors when I try to compile for RISC-V on current main branch (`e16c5bf2e43c8c09e2947c466d9c85eafbce3ec5`). To reproduce:
1) Create a toolchain file `cmake/toolchain/gcc.riscv.cmake
` (assume you have any riscv-gcc installed):
```
set(CMAKE_SYSTEM_NAME Linux     )
set(CMAKE_SYSTEM_PROCESSOR riscv64  )

set(CMAKE_C_COMPILER     riscv64-linux-gnu-gcc  )
set(CMAKE_CXX_COMPILER   riscv64-linux-gnu-g++  )
set(CMAKE_CXX_FLAGS     "-O3 -march=rv64gcv -std=c++20 -W ${EVE_OPTIONS}" )
```
2) Run config:
```
cmake -B build \
  -S . -G Ninja\
  -DCMAKE_TOOLCHAIN_FILE=cmake/toolchain/gcc.riscv.cmake
```

3) Run build (for example, of `unit.arch`):
```
cmake --build build --parallel 4 --target unit.arch.exe
```

On this set up, I see a lot of errors, the first one is:
```
.../include/eve/forward.hpp:42:80: error: no match for 'operator>=' (operand types are 'const eve::cpu_' and 'const eve::avx512_')
   42 | auto EVE_FORCEINLINE to_logical(C c, eve::as<T>) noexcept requires(current_api >= avx512);
      |                                                                    ~~~~~~~~~~~ ^~ ~~~~~~
      |                                                                    |              |
      |                                                                    |              const eve::avx512_
      |                                                                    const eve::cpu_
```


> See how predefs are done for arm-sve:

Yep, I suppose now they are working the same for RISC-V.